### DISCUSSION TITLE
Developing for pipx install in light of PEP 751 (lock files again)

### DISCUSSION THREAD

**Discussion by itcarroll** [pypa/pipx#1548] [Q&A]
I would like to know whether [PEP 751](https://peps.python.org/pep-0751/) is being considered by pipx developers and how pipx might use the proposed "pylock.toml" file.

The PEP rational begins "The format is designed so that a locker which produces the lock file and an installer which consumes the lock file can be separate tools." Because [package locking](https://peps.python.org/pep-0751/#package-locking) is a need for Python applications and pipx is a (the?) primary PyPA project for installing Python applications, I guess pipx should be a consumer of lock files. Is that planned?

One thing I'm not clear on with the PEP is whether "pylock.toml" is meant to be part of an application package (i.e. in the distribution on pypi). Is that the idea? How else would pipx get to consume it?

**Comment by guillermodotn** [comment]
The `pylock.toml` file is meant for **environment management** rather than **packaging**. As such, I don't see it being included in tools like `pipx`, as this file isn't part of Python packages themselves. It's more relevant during the development phase.

That said, other PyPA projects like [`virtualenv`](https://github.com/pypa/virtualenv) would really benefit from this proposal, as they focus on environment management. 

Regarding the possibility of future inclusion, it might make sense if the standard were adopted for scenarios where you install a package directly from a version control system (VCS) URL, with a command like:

```bash
pipx install VCS_URL
```

This could provide a way for `pipx` to consume the lock file in the context of a development environment setup.

**Reply by itcarroll** [reply]
Now that PEP 751 is accepted, I think there will be some more interest in this discussion!

I think there is [general agreement](https://discuss.python.org/t/the-purpose-of-a-lock-file/38756) that lock files are useful for environment management during **deployment**, in addition to development or other times one needs reproducibility. That is why I asked about `pipx` having some way to use a `pylock.toml` to install an application; how does it not make sense for an application installer to want `pylock.toml` distributed with application packages?

**Reply by guillermodotn** [reply]
You're right; to be honest, I don't know what I was thinking when I wrote that. It makes total sense for pipx to support `pylock.toml` when creating isolated environments.

I also expect that `pylock.toml` will be bundled with application packages for distribution.

As for the implementation, it's unclear for me whether `pipx` will handle it directly or if it will rely on the underlying `pip` tooling.

**Comment by itcarroll** [comment]
Anyone stumbling on this might want to jump into a discussion about lockfiles in wheels.

https://discuss.python.org/t/pre-pep-add-ability-to-install-a-package-with-reproducible-dependencies/99497
